mcwb/volume.py

```diff
@@ -0,0 +1,131 @@
+    size. The starting point can be any vertex or the middle of one of the
+    horizontal faces, using the cardinal terminology defined in Anchor3.
+    """
+
+    def __init__(
+        self,
+        position: Vec3,
+        size: Vec3 = None,
+        anchor: Anchor3 = Anchor3.BOTTOM_NW,
+        end: Vec3 = None,  # opposite corner instead of size and anchor
+    ):
+        if (size is None) == (end is None):
+            raise InvalidArgument(
+                "Volume constructor takes only one of size or end")
+        position = Vec3(*position)
+
+        if end is None:
+            size = Vec3(*size)
+            offset = Vec3(0, 0, 0)
+            if anchor.value in Anchor3.TOP.value:
+                offset += Vec3(0, 1 - size.y, 0)
+            if anchor.value in Anchor3.SOUTH.value:
+                offset += Vec3(0, 0, 1 - size.z)
+            if anchor.value in Anchor3.EAST.value:
+                offset += Vec3(1 - size.x, 0, 0)
+            if anchor.value in Anchor3.MIDDLE_FACE.value:
+                offset -= Vec3(int(size.x / 2), 0, int(size.z / 2))
+            elif anchor.value == Anchor3.MIDDLE.value:
+                offset = Vec3(-1, -1, -1) * (size / 2).with_ints()
+
+            self.start = position + offset
+            self.end = self.start + (size - 1)
+        else:
+            end = Vec3(*end)
+            # normalize start(position) and end so all start coords are minima
+            self.start = Vec3(
+                min(position.x, end.x), min(
+                    position.y, end.y), min(position.z, end.z)
+            )
+            self.end = Vec3(
+                max(position.x, end.x), max(
+                    position.y, end.y), max(position.z, end.z)
+            )
+            position = self.start
+            anchor = Anchor3.BOTTOM_NW
+            size = (self.end - position) + 1
+
+        self.size = size
+        self.position = position
+        self.anchor = anchor
+
+    def inside(
+        self, position: Vec3, xtol: int = 0, ytol: int = 0, ztol: int = 0
+    ) -> bool:
+        """ determine if position is within the Volume """
+        return (
+            self.start.x - xtol <= position.x <= self.end.x + xtol
+            and self.start.y - ytol <= position.y <= self.end.y + ytol
+            and self.start.z - ztol <= position.z <= self.end.z + ztol
+        )
+
+    def move(self, distance: Vec3) -> None:
+        """ move the volume's location in space by distance """
+        self.start += distance
+        self.position += distance
+
+    def move_to(self, position: Vec3) -> None:
+        """ move the volume's location in space to position """
+        self.start += position - self.position
+        self.position = position
+
+    def fill(self, client: Client, block: Item = Item.AIR):
+        """ Fill the Volume with a single block type, supports large volumes """
+        if self.size.volume < 32768:
```

Comment:
Where does this magic number for the volume come from?
Why does it matter for the branch condition?

Reply:
The magic number is the maximum number of blocks that the server command fill will accept. I have defines a constant called MAX_MINECRAFT_FILL_COMMAND instead of using magic.